src/fsharp/LexFilter.fs

```diff
@@ -1676,7 +1676,11 @@ type LexFilterImpl (lightSyntaxStatus:LightSyntaxStatus, compilingFsLib, lexer,
                 
         //  module ... ~~~> CtxtModuleHead 
         |  MODULE,(_ :: _) -> 
-            insertComingSoonTokens("MODULE", MODULE_COMING_SOON, MODULE_IS_HERE)
+            match peekNextToken() with
+            | COLON ->
+                insertComingSoonTokens("MODULE", MODULE_ATTRIBUTE_TARGET_COMING_SOON, MODULE_ATTRIBUTE_TARGET_IS_HERE)
```

Comment:
Was there a specific reason that this token-insertion approach needed to fix the bug?  Although the technique occurs elsewhere in pars.fsy/LexFilter.fs, I didn't write that code - another contributor decided to use the technique and it seemed to do the job, so was accepted.

Could you add a comment here saying why the technique is needed?, e.g. an example that doesn't work without it.

Reply:
That change was required to address this:
```
[< module Test =
```
Yacc immediately assumes we are working with an attribute target, and everything falls apart from there.

If there is a better way to do this, I'm all for it. After reading and learning about Yacc and what its limitations are, I don't think there is another way. We have to create these lookahead tokens.

The main problem is Yacc. It's limited; there is no backtracking or peeking/lookahead. The problem not having those capabilities is that it's impossible to express this:

`[< module` or `[< module Test` - This means we are putting an attribute on a module.
`[< module :` - This means we are using an attribute target, because of the "colon".

In Yacc, say we just read the `[<` token, it knows we are going to be parsing an attribute. There is no way to peek ahead to see if there is a `module :` there. This is impossible and indeed not how Yacc works at all. As soon as `module` has been read, it's already over.

In order to help mitigate these limitations, the way to handle looking ahead is to provide lookahead tokens before running Yacc. In our case, we call them COMING_SOON tokens. Our LexFilter does just that. Arbitrarily, it inserts 6 coming soon tokens before the real token, and actually replaces the real token with a IS_HERE token to prevent it re-inserting COMING_SOON tokens in an infinite loop. Most of this is documented in the LexFilter, except for the IS_HERE bit.

So, when the LexFilter reads: 
`MODULE`
It then turns into:
`MODULE_COMING_SOON MODULE_COMING_SOON MODULE_COMING_SOON MODULE_COMING SOON MODULE_COMING_SOON MODULE_COMING_SOON MODULE_IS_HERE`

So, to handle lookahead for one token in Yacc, you can now do that against modules and pop off one coming soon token at a time.

In the our case:
```
[< module Test =
```
We need to peek ahead twice instead of once. Therefore, we need a COMING_SOON token that represents that two specific tokens are coming soon. Which is `MODULE_ATTRIBUTE_TARGET_COMING_SOON`.

Reply:
OK.  I understand.

Yes, the technique is ok.  Please add a condensed version of the above to the places where COMING_SOON tokens are inserted.